### PR TITLE
mapAt

### PR DESCRIPTION
**Pull request by theryee**
Issue #85

### CODE REVIEW

**Comment by theryee** [comment]
Does it make sense to add at() to other map-like structures in the EASTL?

File: include/EASTL/map.h [hunk 1]
```diff
@@ -11,6 +11,7 @@
 #include <EASTL/internal/red_black_tree.h>
 #include <EASTL/functional.h>
 #include <EASTL/utility.h>
+#include <stdexcept>
```
Comment (rparolin) [inline include/EASTL/map.h]: This shouldn't be necessary as it should be added in the config.h header.

File: include/EASTL/map.h [hunk 2]
```diff
@@ -439,6 +443,33 @@ namespace eastl
...
 
 
+	template <typename Key, typename T, typename Compare, typename Allocator>
+	inline T& map<Key, T, Compare, Allocator>::at(const Key& key)
+	{
+		iterator itLower(lower_bound(key)); // itLower->first is >= key.
+
+		if(itLower == end())
+		{
+			throw std::out_of_range("map::at key does not exist");
```
Comment (rparolin) [inline include/EASTL/map.h]: EASTL can't assume that exceptions are enabled.  The vast majority of the time we disable exceptions so we typical add asserts in the case when exceptions are disabled.  You can see examples of how this done through the library but most recently I implemented string_view with similar code.

https://github.com/electronicarts/EASTL/blob/master/include/EASTL/string_view.h#L89

File: test/packages/EATest/include/EATest/EATest.h [hunk 3]
```diff
@@ -341,6 +341,17 @@ namespace EA
             #define EATEST_VERIFY_F(bExpression, pFormat, ...) EA::UnitTest::TestInternal::EATEST_VERIFY_F_IMP((bExpression), nErrorCount, __FILE__, __LINE__, pFormat, __VA_ARGS__)
         #endif
 
+				/// EATEST_VERIFY_THROW
+				/// EATEST_VERIFY_NOTHROW
+				/// 
+				/// This macro confirms whether or not an expression throws or doesn't throw.
+				/// 
+				/// See EATEST_VERIFY for details about error reporting and the _MSG variants
+				#define EATEST_VERIFY_THROW(expression) {bool isNoThrow; try{ {expression;} isNoThrow=true; }catch(...){isNoThrow=false;} EATEST_VERIFY(!isNoThrow); }
```
Comment (rparolin) [inline test/packages/EATest/include/EATest/EATest.h]: Similarly code that assumes to throw must be wrapped with exception preprocessor guards.

File: test/source/TestMap.cpp [hunk 4]
```diff
@@ -141,6 +140,23 @@ int TestMap()
 		EATEST_VERIFY(v.validate());
 	}
 
+	{
+		typedef eastl::map<int, int>     IntIntMap;
+		IntIntMap map1;
+
+		EATEST_VERIFY_THROW(map1.at(0));
+		map1[0]=1;
+		EATEST_VERIFY_NOTHROW(map1.at(0));
```
Comment (rparolin) [inline test/source/TestMap.cpp]: Again preprocessor guards for exceptions.

File: include/EASTL/map.h (hunk 1 above)
Comment (theryee) [inline include/EASTL/map.h]: I'll trace this out. I was getting missing declaration errors on linux gcc.

File: include/EASTL/map.h (hunk 2 above)
Comment (theryee) [inline include/EASTL/map.h]: k

File: test/packages/EATest/include/EATest/EATest.h (hunk 3 above)
Comment (theryee) [inline test/packages/EATest/include/EATest/EATest.h]: k

File: test/source/TestMap.cpp (hunk 4 above)
Comment (theryee) [inline test/source/TestMap.cpp]: k

File: include/EASTL/map.h [hunk 5]
```diff
@@ -439,6 +442,47 @@ namespace eastl
 	#endif
 
 
+	template <typename Key, typename T, typename Compare, typename Allocator>
+	inline T& map<Key, T, Compare, Allocator>::at(const Key& key)
+	{
+		iterator itLower(lower_bound(key)); // itLower->first is >= key.
+
+		#if EASTL_EXCEPTIONS_ENABLED
+			if(itLower == end())
```
Comment (rparolin) [inline include/EASTL/map.h]: Since the branch is checked in both cases, I'd promote it outside of the preprocessor guards.

Comment (rparolin) [inline include/EASTL/map.h]: Other than that, I think its ready to go.

**Comment by rparolin** [comment]
Awesome!  Thanks for the PR.

**Pull request merged**